### ISSUE TITLE
Is flash attention 3 compatible with Ulysses or not?

### ISSUE THREAD

**Issue by kisseternity** [Dao-AILab/flash-attention#1834]
Hello, I've been training Qwen3 recently with FA3 and Ulysses sp > 1, but I get nan logits after forward:
(WorkerDict pid=8874) WARN: rank 1 non-finite policy loss. pg_loss_finite=False ppo_kl_finite=False. old_log_prob_stats={'shape': (3, 4096), 'nan': 136, 'inf': 0, 'min': -26.433170318603516, 'max': 0.0} log_prob_stats={'shape': (3, 4096), 'nan': 0, 'inf': 0, 'min': 

The forward code:
                # gather log_prob if sp > 1
                if self.use_ulysses_sp:
                    # gather and unpad for the ulysses sp
                    log_probs = gather_outputs_and_unpad(
                        log_probs,
                        gather_dim=0,
                        unpad_dim=0,
                        padding_size=pad_size,
                    )
                    if calculate_entropy:
                        entropy_rmpad = gather_outputs_and_unpad(
                            entropy_rmpad,
                            gather_dim=0,
                            unpad_dim=0,
                            padding_size=pad_size,
                        )
                # pad back to (bsz, seqlen)
                if calculate_entropy:
                    full_entropy = pad_input(
                        hidden_states=entropy_rmpad.unsqueeze(-1),
                        indices=indices,
                        batch=batch_size,
                        seqlen=seqlen,
                    )
                full_log_probs = pad_input(
                    hidden_states=log_probs.unsqueeze(-1),
                    indices=indices,
                    batch=batch_size,
                    seqlen=seqlen,
                )

                # only return response part:
                if calculate_entropy:
                    entropy = full_entropy.squeeze(-1)[:, -response_length - 1 : -1]  # (bsz, response_length)
                log_probs = full_log_probs.squeeze(-1)[:, -response_length - 1 : -1]  # (bsz, response_length)

            else:  # not using rmpad and no ulysses sp
                extra_args = {}
                if self.use_fused_kernels:
                    extra_args["temperature"] = temperature
                    extra_args["return_dict"] = True

                output = self.actor_module(
                    input_ids=input_ids,
                    attention_mask=attention_mask,
                    position_ids=position_ids,
                    **multi_modal_inputs,
                    use_cache=False,
                    **extra_args,
                )  # prevent model thinks we are generating

                if self.use_fused_kernels:
                    log_probs = output.log_probs[:, -response_length - 1 : -1]
                    entropy = output.entropy[:, -response_length - 1 : -1]  # (bsz, response_length)

                else:
                    logits = output.logits

                    logits.div_(temperature)
                    logits = logits[:, -response_length - 1 : -1, :]  # (bsz, response_length, vocab_size)
                    log_probs = logprobs_from_logits(logits, micro_batch["responses"])
                    if calculate_entropy:
                        if not self.config.entropy_checkpointing:
                            entropy = verl_F.entropy_from_logits(logits)  # (bsz, response_length)
                        else:
                            entropy = torch.utils.checkpoint.checkpoint(verl_F.entropy_from_logits, logits)
            try:
                rank = torch.distributed.get_rank() if torch.distributed.is_initialized() else -1
            except Exception:
                rank = -1
            def _tensor_stats(t: torch.Tensor):
                if t is None:
                    return None
                t_float = t.float()
                num_nan = torch.isnan(t_float).sum().item()
                num_inf = torch.isinf(t_float).sum().item()
                return {
                    "shape": tuple(t.shape),
                    "nan": int(num_nan),
                    "inf": int(num_inf),
                    "min": float(torch.nan_to_num(t_float, nan=0.0, posinf=0.0, neginf=0.0).min().item()),
                    "max": float(torch.nan_to_num(t_float, nan=0.0, posinf=0.0, neginf=0.0).max().item()),
                }
            has_non_finite = (not torch.isfinite(log_probs).all()) or (
                (entropy is not None) and (not torch.isfinite(entropy).all())
            )
            if has_non_finite:
                print(
                    f"WARN: rank {rank} non-finite in actor forward. "
                    f"use_remove_padding={self.use_remove_padding} use_fused_kernels={self.use_fused_kernels} "
                    f"temperature={temperature}. log_probs_stats={_tensor_stats(log_probs)} "
                    f"entropy_stats={_tensor_stats(entropy)}"
                )

I wonder if fa3 is compatible with Ulysses sp or not? Thanks.
